expo/expo
# [docs] Limit preval usage and simplify caching

# Why

This PR does 3 things to help with the usage and performance of the docs. Right now, we use two different caches in CI to separate MDX content and Next + Preval content. With this change, we separate the caches with MDX + Next content and Preval content.

1. Preval has [issues invalidating the cache](https://github.com/kentcdodds/babel-plugin-preval/issues/19). Using this in more places than absolute necessary might mess up the cache more. This change removes the "global" Preval plugin. Instead, it adds a special loader for all `docs/constants/**` files and includes Preval within those files.

2. Because we consolidated Next and MDX caches, we can simplify the CI. With this change we only have a single Next cache, and never store prevaled files.

3. We also had an issue where a single `yarn export` run created two different build ids, causing some issues with the JS/HTML again. It could be related to us caching the full `.next` folder, [instead of just `.next/cache`](https://github.com/vercel/next.js/blob/canary/errors/no-cache.md#possible-ways-to-fix-it). This changes to only `.next/cache`, removing any traces of previous builds.

I couldn't split this into 3 different PRs unfortunately, they are pretty much all related.

# How

- Moved the constant files in `common/` to `constants/` and replaced all occurrences
- Limited Preval usage to only `constants/` files by adding a custom webpack loader in `next.config.js`
- [Traced the last "JS not found" error to multiple build IDs being published](https://github.com/expo/expo/runs/1063326056?check_suite_focus=true#step:12:112)


# Test Plan

**To test the Preval cache**
- `$ yarn next dev` to build first cache
- Remove random pages from `pages/versions/v38.0.0/sdk`
- Validate there are no changes in the sidebar (on any `v38.0.0` page), resulting in `404` from the removed pages
- Manually remove `.next/preval` folder, and restart `$ yarn next dev` (will remove all other files/folders except `cache` in `.next`)
- Validate there are changes in the sidebar, all removed pages should be gone

**To test the cache**
- Validate if CI only caches `.next/cache`
- There should be no double-id-export occurrences
- Check if there is no warning about ["no build cache"](https://github.com/vercel/next.js/blob/canary/errors/no-cache.md)
